File: cepi-ios/CEPITelemedicina/Galeria/GaleriaModelo.swift
import Foundation
import Observation

/// Las imágenes clínicas que se ven en una rejilla: la galería de la organización o las de un
/// paciente (PAPER §24.2.1). Pide de a páginas y espera un momento antes de buscar, para no
/// mandar una consulta por tecla.
@MainActor
@Observable
final class GaleriaModelo {
    /// Solo las de este paciente; `nil` = todas las de la organización activa.
    let paciente: String?

    private(set) var imagenes: [ImagenGaleria] = []
    private(set) var total: Int?
    private(set) var cargando = false
    private(set) var cargado = false
    private(set) var error: String?

    @ObservationIgnored private var pedido = 0
    @ObservationIgnored private var busquedaAplicada = ""

    private static let porPagina = 60

    init(paciente: String? = nil) {
        self.paciente = paciente
    }

    var hayMas: Bool {
        guard let total else { return false }
        return imagenes.count < total
    }

    /// Texto que se está mostrando; cambia solo cuando la búsqueda ya se aplicó.
    var busqueda: String { busquedaAplicada }

    /// Llamado desde `.task(id: busqueda)`: espera un momento (la tecla siguiente cancela la
    /// tarea) y recarga desde cero.
    func buscar(_ texto: String, api: CEPIAPI, esperar: Duration = .milliseconds(350)) async {
        let limpio = texto.trimmingCharacters(in: .whitespaces)
        if limpio != busquedaAplicada, !limpio.isEmpty || cargado {
            try? await Task.sleep(for: esperar)
            guard !Task.isCancelled else { return }
        }
        busquedaAplicada = limpio
        await cargar(api: api, desde: 0)
    }

    func siguientePagina(api: CEPIAPI) async {
        guard hayMas, !cargando else { return }
        await cargar(api: api, desde: imagenes.count)
    }

    func recargar(api: CEPIAPI) async {
        await cargar(api: api, desde: 0)
    }

    private func cargar(api: CEPIAPI, desde: Int) async {
        pedido += 1
        let mio = pedido
        cargando = true
        defer { if mio == pedido { cargando = false } }
        do {
            let pagina = try await api.galeria(
                texto: busquedaAplicada, paciente: paciente, limite: Self.porPagina, desde: desde
            )
            guard mio == pedido else { return }   // llegó otra búsqueda más nueva
            imagenes = desde == 0 ? pagina.data : imagenes + pagina.data
            total = pagina.total
            error = nil
            cargado = true
        } catch {
            guard mio == pedido, !Task.isCancelled else { return }
            // Un fallo al pedir "más" no borra lo que ya se está viendo.
            if desde == 0 { imagenes = [] }
            self.error = error.localizedDescription
        }
    }
}

File: cepi-android/app/src/main/java/ec/cepi/telemedicina/galeria/GaleriaModelo.kt
package ec.cepi.telemedicina.galeria

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import ec.cepi.telemedicina.api.ApiError
import ec.cepi.telemedicina.api.CepiApi
import ec.cepi.telemedicina.api.ImagenGaleria
import kotlinx.coroutines.delay

/**
 * Las imágenes clínicas que se ven en una rejilla: la galería de la organización o las de un
 * paciente (PAPER §24.2.1). Pide de a páginas y espera un momento antes de buscar, para no
 * mandar una consulta por tecla. Espejo de `GaleriaModelo.swift`.
 */
class GaleriaModelo(
    private val api: CepiApi,
    /** Solo las de este paciente; `null` = todas las de la organización activa. */
    val paciente: String? = null,
) {
    var imagenes: List<ImagenGaleria> by mutableStateOf(emptyList())
        private set
    var total: Int? by mutableStateOf(null)
        private set
    var cargando: Boolean by mutableStateOf(false)
        private set
    var cargado: Boolean by mutableStateOf(false)
        private set
    var error: String? by mutableStateOf(null)
        private set
    /** Texto que se está mostrando; cambia solo cuando la búsqueda ya se aplicó. */
    var busqueda: String by mutableStateOf("")
        private set

    private var pedido = 0

    val hayMas: Boolean get() = total?.let { imagenes.size < it } ?: false

    /**
     * Llamado desde un `LaunchedEffect(texto)`: espera un momento (la tecla siguiente cancela la
     * corrutina) y recarga desde cero.
     */
    suspend fun buscar(texto: String, esperaMs: Long = 350) {
        val limpio = texto.trim()
        if (limpio != busqueda && (limpio.isNotEmpty() || cargado)) delay(esperaMs)
        busqueda = limpio
        cargar(desde = 0)
    }

    suspend fun siguientePagina() {
        if (!hayMas || cargando) return
        cargar(desde = imagenes.size)
    }

    suspend fun recargar() = cargar(desde = 0)

    private suspend fun cargar(desde: Int) {
        val mio = ++pedido
        cargando = true
        try {
            val pagina = api.galeria(busqueda, paciente, POR_PAGINA, desde)
            if (mio != pedido) return   // llegó otra búsqueda más nueva
            imagenes = if (desde == 0) pagina.data else imagenes + pagina.data
            total = pagina.total
            error = null
            cargado = true
        } catch (e: ApiError) {
            if (mio != pedido) return
            // Un fallo al pedir "más" no borra lo que ya se está viendo.
            if (desde == 0) imagenes = emptyList()
            error = e.mensaje
        } finally {
            if (mio == pedido) cargando = false
        }
    }

    private companion object {
        const val POR_PAGINA = 60
    }
}
